libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,719 @@
+/*
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "mqtt_state.h"
+
+#define MQTT_PACKET_ID_INVALID ( uint16_t ) 0
+
+/**
+ * @brief Set the appropriate state when the packet is one of PUBACK,
+ * PUBREC, PUBREL, or PUBCOMP.
+ * 
+ * @param[in] opType Reserve, Send, or Receive.
+ * @param[in] idStatus Invalid, Unknown, or Known.
```

Comment:
I think there needs to be separation of concerns.  The reserve operation is strictly associated with adding an outbound record, so it need not be present in all these mapping functions.  Likewise, the only legal case for an unknown (barring the trivial reserve) is a receive operation, so again the mapping functions can be simplified further.  Since what I'm suggesting means quite a few changes, let's talk it over together.